android/src/main/java/com/tailscale/ipn/ui/notifier/Notifier.kt

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) 2024 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package com.tailscale.ipn.ui.notifier
+
+import android.util.Log
+import com.tailscale.ipn.ui.model.Ipn.Notify
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.serialization.decodeFromString
+import kotlinx.serialization.json.Json
+
+typealias NotifierCallback = (Notify) -> Unit
+
+
+class Watcher(
+    val sessionId: String,
+    val mask: Int,
+    val callback: NotifierCallback
+)
+
+// Notifier is a wrapper around the IPN Bus notifier.  It provides a way to watch the
+// for changes in various parts of the Tailscale engine.  You will typically only use
+// a single Notifier per instance of your application which lasts for the lifetime of
+// the process.
+//
+// The primary entry point here is watchIPNBus which will start a watcher on the IPN bus
+// and return you the session Id.  When you are done with your watcher, you must call
+// unwatchIPNBus with the sessionId.
+class Notifier {
```

Comment:
make this a singleton?

Reply:
It is a defacto singleton because it's a property on a singleton... But for testing, I want to be able to inject these.